src/tools/scripts/update_timestamps_versions_on_commit.sh

```diff
@@ -0,0 +1,97 @@
+#!/bin/bash
+
+######################################
+## Custom Web Almanac script        ##
+######################################
+#
+# This script updates CSS and JS reference to new versions numbers
+# when they are included in a commit. This is used for cahce busting.
+# 
+# It also updates timestamps for any .md and .html files changed
+#
+# This script is run in a GitHub Action and should not need to be run manually
+#
+# Note:  we use basic sed with no support for "+" (one or more) and instead
+# fake it with syntax like "[0-9][0-9]*" to avoid having to enable extended
+# regex which would require more escapes.
+#
+
+if [ "$#" -eq 1 ]; then
+COMMIT_SHA=$1
+fi
+
+if [ "{COMMIT_SHA}" = "" ]; then
+    echo "Must supply commit sha"
+    exit 1
+fi
+ 
+
+# This script must be run from src directory
+if [ -d "src" ]; then
+  cd src
+fi
+
+if [ $(uname) = 'Darwin' ]; then
+  echo "Running MacOS"
+  MACOS=true
+fi
+
+NEW_LONG_DATE=`date +%Y%m%d%H%M%S`
+NEW_SHORT_DATE=`date '+%Y-%m-%d'`
+
+echo "Files changed in this commit:"
+git diff-tree --diff-filter=AM --no-commit-id --name-only -r ${COMMIT_SHA}
+
+function update_versions {
+  FILE_TYPE=$1
+  echo "Updating ${FILE_TYPE} versions"
+  CHANGED_FILES=`git diff-tree --diff-filter=AM --no-commit-id --name-only -r ${COMMIT_SHA} static/${FILE_TYPE} | grep "\.${FILE_TYPE}"`
+  for CHANGED_FILE in ${CHANGED_FILES}
+  do
+    CHANGED_FILE=`basename ${CHANGED_FILE}`
+    echo "Updating ${CHANGED_FILE} version number to ${NEW_LONG_DATE}"
+    if [ "${MACOS}" = true ]; then
+      sed -i '' s/${CHANGED_FILE}\?v=[0-9][0-9]*/${CHANGED_FILE}\?v=${NEW_LONG_DATE}/ templates/base/*/*.html
+      sed -i '' s/${CHANGED_FILE}\?v=[0-9][0-9]*/${CHANGED_FILE}\?v=${NEW_LONG_DATE}/ templates/base.html
+    else
+      sed -i s/${CHANGED_FILE}\?v=[0-9][0-9]*/${CHANGED_FILE}\?v=${NEW_LONG_DATE}/ templates/base/*/*.html
+      sed -i s/${CHANGED_FILE}\?v=[0-9][0-9]*/${CHANGED_FILE}\?v=${NEW_LONG_DATE}/ templates/base.html
+    fi
```

Comment:
Just thinking out loud, to avoid this OS branching, should we define a variable in the beginning of the script with the name `SED_I_FLAG` and set its value to `-i` or `-i ''` depending on the OS. Then use this variable when composing the command, this way, we will avoid many nested if/else branches here and in the `update_timestamp` function.

Reply:
I tried that and failed and got frustated and gave up. Also if we do use extended regex (and not 100% convinced on that!) then MacOS has different way of invoking this than (`sed -E` as opposed to `sed -r` for linux) so might be another reason to keep two commands.

Reply:
```
$ SED_FLAGS="-i -r"
$ echo sed $SED_FLAGS 's/foo/bar/' baz
sed -i -r s/foo/bar/ baz
$ SED_FLAGS="-i '' -E"
$ echo sed $SED_FLAGS 's/foo/bar/' baz
sed -i '' -E s/foo/bar/ baz
```

Reply:
And now it works first time when I try again. Sighhh... Must have been tired when I did this late last night. Resolved.

Reply:
Well, empty strings could be tricky to deal with in shells, so I guess you had some quoting and string interpolation issues.

Reply:
Nope. It just doesn't work on MacOS:

```sh
$ cat test.html
barry

$ cat test.sh
#!/bin/bash

if [ $(uname) = 'Darwin' ]; then
  echo "Running MacOS"
  SED_FLAGS="-i '' -E"
else
  SED_FLAGS="-i -r"
fi

echo "sed ${SED_FLAGS} /barry/pollard/"
sed ${SED_FLAGS} s/barry/pollard/g test.html

$ ./test.sh
Running MacOS
sed -i '' -E /barry/pollard/

$ ls -la test*
-rw-r--r--  1 barry  staff    8  6 Aug 18:54 test.html
-rw-r--r--  1 barry  staff    6  6 Aug 18:53 test.html''
-rwxr-xr-x  1 barry  staff  207  6 Aug 18:54 test.sh
```

I dunno why it creates that `test.html''` file and puts changes in there!!!

Eventually [got this to work](https://stackoverflow.com/questions/5694228/sed-in-place-flag-that-works-both-on-mac-bsd-and-linux):

```sh
# Default case for Linux sed, just use "-i -r"
SED_FLAGS=(-i -r)
if [ $(uname) = 'Darwin' ]; then
  echo "Running MacOS"
  SED_FLAGS=(-i "" -E)
fi

...

sed "${SED_FLAGS[@]}" "s/...
```

Very frustrating!!!

Reply:
Yes, I guess single quotes were internally escaped when places inside of double quotes. However, using parenthesis caused them to be a list of token that is then joined in place.